runtime/doc/lua-plugin.txt

```diff
@@ -0,0 +1,290 @@
+      for i = #lines, 1, -1 do
+        if lines[i]:find(target_pattern) == nil then
+          start_line = i
+          break
+        end
+      end
+      start_line = start_line or 0
+
+      -- Remove trailing lines
+      vim.api.nvim_buf_set_lines(buf_id, start_line, -1, false, {})
+    end
+
+    -- Return the module. This must be the last of the file.
+    return M
+<
+And this is it. You have create a Lua plugin. Users now can install it
+(i.e. put the whole `no-trailing/` directory in their 'runtimepath' manually
+or via plugin manager) and use it directly with `require()`:
+
+• Any of `require('no-trailing').trim(0)`, `require('no-trailing').trim(0, {})`,
+  or `require('no-trailing').trim(0, { empty = false })` removes all
+  blank trailing lines.
+
+• `require('no-trailing').trim(0, { empty = true })` removes only empty
+  trailing lines.
+
+• Any of `require('no-trailing').trim()`, `require('no-trailing').setup('a')`,
+  `require('no-trailing').trim(0, { empty = 'a' })` will throw an error.
+
+Notes:
+
+• Choosing function scope (i.e. what it does) is up to you. Usually preferring
+  the functionality just enough for it to be useful for target audience is
+  a good approach. You can see what Nvim does in |dev-lua|.
+
+• Choosing names for plugin/function/arguments/options is up to you. Usually
+  choosing a short "to the point" yet descriptive name is a good approach.
+  You can see what Nvim does in |dev-naming|.
+
+• Choosing function signature (i.e its arguments) is up to you. Preferring
+  first mandatory arguments followed by a single optional table `opts` with
+  optional arguments is a common convention.
+  You can see what Nvim uses in |dev-patterns| and |dev-api|.
+
+• To interactively test latest plugin functionality, open separate fresh Nvim
+  instance with plugin loaded and interact with it.
+
+------------------------------------------------------------------------------
+Restructuring Lua code                                  *lua-plugin-restructure*
+
+Using single Lua file in plugin is possible, yet may become not easy to manage
+if it becomes too big. To overcome this, code can be split into separate
+modules and used inside each other with |require()|.
+
+In our 'no-trailing' plugin computation of the first trailing line can be
+moved into a separate `lua/no-trailing/utils.lua` file:
+>lua
+    -- lua/no-trailing/utils.lua
+
+    local M = {}
+
+    M.get_last_trailing = function(buf_id, empty)
+      local target_pattern = empty and '^$' or '^%s*$'
+      local lines = vim.api.nvim_buf_get_lines(buf_id, 0, -1, false)
+      for i = #lines, 1, -1 do
+        if lines[i]:find(target_pattern) == nil then
+          -- Moving computation into a separate function allows using the
+          -- "early return" pattern
+          return i
+        end
+      end
+      return 0
+    end
+
+    return M
+<
+
+Now there are two main approaches on how this can be used in main module:
+eager loading and lazy loading.
+
```

Comment:
There's a LOT of code to wade through here even though the relevant part is only like 3 lines. I wonder if the "build plugins from scratch" is the way to go here, it leads to *lot* of code reading.

Reply:
+100. See also https://github.com/neovim/neovim/issues/22366#issuecomment-1445501976